modAionImpl/src/org/aion/zero/impl/db/PendingBlockStore.java

```diff
@@ -0,0 +1,816 @@
+                blocks.put(ByteArrayWrapper.wrap(queue), list);
+            }
+        }
+
+        return blocks;
+    }
+
+    /**
+     * Used to delete imported queues from storage.
+     *
+     * @param level the block height of the queue starting point
+     * @param queues the identifiers for the queues to be deleted
+     * @param blocks the queue to blocks mappings to me deleted (used to ensure that if the queues
+     *     have been expanded, only the relevant blocks get deleted)
+     */
+    public void dropPendingQueues(
+            long level,
+            Collection<ByteArrayWrapper> queues,
+            Map<ByteArrayWrapper, List<AionBlock>> blocks) {
+
+        lock.writeLock().lock();
+
+        try {
+            // delete imported queues & blocks
+            for (ByteArrayWrapper q : queues) {
+                // load the queue from disk
+                List<AionBlock> currentQ = queueSource.get(q.getData());
+
+                // delete imported blocks
+                for (AionBlock b : blocks.get(q)) {
+                    // delete index
+                    indexSource.putToBatch(b.getHash(), null);
+                    currentQ.remove(b);
+                }
+
+                // delete queue
+                queueSource.putToBatch(q.getData(), null);
+
+                // the queue has been updated since the import read
+                if (!currentQ.isEmpty()) {
+                    // get first block in remaining queue
+                    AionBlock first = currentQ.get(0);
+
+                    // update queue hash to first remaining element
+                    byte[] currentQueueHash = first.getHash();
+
+                    // put in queue database
+                    queueSource.putToBatch(currentQueueHash, currentQ);
+
+                    // update block index
+                    for (AionBlock b : currentQ) {
+                        indexSource.putToBatch(b.getHash(), currentQueueHash);
+                    }
+
+                    // add (to) level
+                    byte[] levelKey = ByteUtil.longToBytes(first.getNumber());
+                    List<byte[]> levelData = levelSource.get(levelKey);
+
+                    if (levelData == null) {
+                        levelData = new ArrayList<>();
+                    }
+
+                    levelData.add(currentQueueHash);
+                    levelSource.putToBatch(levelKey, levelData);
+                }
+            }
+
+            // update level
+            byte[] levelKey = ByteUtil.longToBytes(level);
+            List<byte[]> levelData = levelSource.get(levelKey);
+
+            if (levelData == null) {
+                LOG.error(
+                        "Corrupt data in PendingBlockStorage. Level (expected to exist) was not found.");
+                // level already missing so nothing to do here
+            } else {
+                List<byte[]> updatedLevelData = new ArrayList<>();
+
+                levelData.forEach(
```

Comment:
I don't have strong feelings about it, but I've read that Collection.forEach(<lambda>) is generally slower than doing a plain old for(Object x : X) loop (unless you use `parallelStream()`).  Just wanted to point that out if performance is a big consideration.

Reply:
updated in 2bc2919